projects/ngneat/input-mask/src/lib/input-mask.directive.ts

```diff
@@ -75,19 +84,51 @@ export class InputMaskDirective<T = any>
   }
 
   registerOnChange(fn: (_: T | null) => void): void {
-    const parser = this.inputMask.parser;
-    this.onInput = (value) => {
-      fn(parser ? parser(value) : value);
-    };
+    this.onChange = fn;
+    if (this.getFormControl()?.updateOn === 'change') {
+      const parser = this.inputMask.parser;
+      this.onInput = (value) => {
+        fn(parser ? parser(value) : value);
+      };
+    }
   }
 
-  registerOnTouched(fn: any): void {}
+  registerOnTouched(fn: () => void): void {
+    if (this.getFormControl()?.updateOn === 'blur') {
+      const parser = this.inputMask.parser;
+      this.onInput = (value): void => {
+        const newValue = parser ? parser(value) : value;
+        this.onChange(newValue); // 2 discuss
+        if (this.formInitialValue !== value) {
+          fn();
+        }
+      };
+    }
+  }
 
   validate(): { [key: string]: any } | null {
     return this.inputMaskPlugin && this.inputMaskPlugin.isValid()
       ? null
       : { inputMask: false };
   }
+
+  private getFormControl(): AbstractControl | null {
```

Comment:
Can you please turn this into getter?
```suggestion
  private get formControl(): AbstractControl | null {
```